google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta2/StreamWriterV2.java

```diff
@@ -0,0 +1,364 @@
+
+  /*
+   * Contains sent append requests waiting for response from server.
+   */
+  @GuardedBy("lock")
+  private final Deque<AppendRequestAndResponse> inflightRequestQueue;
+
+  /*
+   * Wraps the underlying bi-directional stream connection with server.
+   */
+  private StreamConnection streamConnection;
+
+  /*
+   * A separate thread to handle actual communication with server.
+   */
+  private Thread appendThread;
+
+  private StreamWriterV2(Builder builder) {
+    this.lock = new ReentrantLock();
+    this.hasMessageInWaitingQueue = lock.newCondition();
+    this.streamName = builder.streamName;
+    this.waitingRequestQueue = new LinkedList<AppendRequestAndResponse>();
+    this.inflightRequestQueue = new LinkedList<AppendRequestAndResponse>();
+    this.streamConnection =
+        new StreamConnection(
+            builder.client,
+            new RequestCallback() {
+              @Override
+              public void run(AppendRowsResponse response) {
+                requestCallback(response);
+              }
+            },
+            new DoneCallback() {
+              @Override
+              public void run(Throwable finalStatus) {
+                doneCallback(finalStatus);
+              }
+            });
+    this.appendThread =
+        new Thread(
+            new Runnable() {
+              @Override
+              public void run() {
+                appendLoop();
+              }
+            });
+    this.appendThread.start();
+  }
+
+  /**
+   * Schedules the writing of a message.
+   *
+   * <p>Example of writing a message.
+   *
+   * <pre>{@code
+   * AppendRowsRequest message;
+   * ApiFuture<AppendRowsResponse> messageIdFuture = writer.append(message);
+   * ApiFutures.addCallback(messageIdFuture, new ApiFutureCallback<AppendRowsResponse>() {
+   *   public void onSuccess(AppendRowsResponse response) {
+   *     if (response.hasOffset()) {
+   *       System.out.println("written with offset: " + response.getOffset());
+   *     } else {
+   *       System.out.println("received an in stream error: " + response.error().toString());
+   *     }
+   *   }
+   *
+   *   public void onFailure(Throwable t) {
+   *     System.out.println("failed to write: " + t);
+   *   }
+   * }, MoreExecutors.directExecutor());
+   * }</pre>
+   *
+   * @param message the message in serialized format to write to BigQuery.
+   * @return the message ID wrapped in a future.
+   */
+  public ApiFuture<AppendRowsResponse> append(AppendRowsRequest message) {
+    AppendRequestAndResponse requestWrapper = new AppendRequestAndResponse(message);
+    this.lock.lock();
+    try {
+      if (userClosed) {
+        requestWrapper.appendResult.setException(
+            new StatusRuntimeException(
+                Status.fromCode(Status.Code.FAILED_PRECONDITION)
+                    .withDescription("Stream is already closed")));
+        return requestWrapper.appendResult;
+      }
+      if (connectionFinalStatus != null) {
```

Comment:
Same here.

Reply:
I don't mind either way, but dataflow prefers us to set the future. So I prefer to keep it this way.